File: app/src/main/java/com/bluetalk/app/bluetooth/FileTransfer.kt
package com.bluetalk.app.bluetooth

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.provider.OpenableColumns
import android.util.Base64
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/** A file stored locally, ready to attach to a message or render. */
data class Attachment(val path: String, val name: String, val mime: String)

/**
 * Moves files between devices over the chat channel. Because a whole image
 * far exceeds one frame, a file is sent as a [Frame.FileStart], a run of
 * ordered [Frame.FileData] slices, and a [Frame.FileEnd]; the receiver
 * streams the slices straight to disk. Outgoing images are downscaled and
 * re-encoded so a photo doesn't take minutes over Bluetooth.
 *
 * Used by both the RFCOMM and BLE managers; each transfer is keyed by a
 * globally unique id so concurrent transfers never interleave.
 */
class FileTransfer(context: Context) {

    private val appContext = context.applicationContext
    private val dir = File(appContext.filesDir, "attachments").apply { mkdirs() }

    private class Partial(
        val name: String,
        val mime: String,
        val file: File,
        val out: BufferedOutputStream,
    )

    private val partials = ConcurrentHashMap<String, Partial>()

    // ---- Receiving ---------------------------------------------------

    fun startIncoming(id: String, name: String, mime: String) {
        val file = File(dir, "$id-${sanitize(name)}")
        partials[id] = Partial(name, mime, file, BufferedOutputStream(FileOutputStream(file)))
    }

    fun appendIncoming(id: String, base64: String) {
        val partial = partials[id] ?: return
        try {
            partial.out.write(Base64.decode(base64, Base64.NO_WRAP))
        } catch (e: IllegalArgumentException) {
            // Corrupt slice; drop it.
        }
    }

    fun finishIncoming(id: String): Attachment? {
        val partial = partials.remove(id) ?: return null
        return try {
            partial.out.flush()
            partial.out.close()
            Attachment(partial.file.absolutePath, partial.name, partial.mime)
        } catch (e: Exception) {
            null
        }
    }

    fun abortIncoming(id: String) {
        partials.remove(id)?.let {
            try {
                it.out.close()
                it.file.delete()
            } catch (e: Exception) {
                // Ignore.
            }
        }
    }

    // ---- Sending -----------------------------------------------------

    /**
     * Copies a picked content Uri into app storage, downscaling and
     * recompressing images. Returns null if the file can't be read.
     */
    fun importOutgoing(uri: Uri): Attachment? {
        val resolver = appContext.contentResolver
        val mime = resolver.getType(uri) ?: "application/octet-stream"
        val name = queryName(uri) ?: "file"
        val id = UUID.randomUUID().toString()
        return try {
            if (mime.startsWith("image/")) {
                val bitmap = resolver.openInputStream(uri).use { BitmapFactory.decodeStream(it) }
                    ?: return null
                val scaled = downscale(bitmap, MAX_IMAGE_DIMENSION)
                val file = File(dir, "$id-${sanitize(name.substringBeforeLast('.'))}.jpg")
                FileOutputStream(file).use { scaled.compress(Bitmap.CompressFormat.JPEG, 80, it) }
                Attachment(file.absolutePath, file.name, "image/jpeg")
            } else {
                val file = File(dir, "$id-${sanitize(name)}")
                resolver.openInputStream(uri).use { input ->
                    FileOutputStream(file).use { output -> input?.copyTo(output) }
                }
                Attachment(file.absolutePath, name, mime)
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Streams a file to a peer as protocol frames, passing each to [send].
     * [send] returns false when the link died, aborting the transfer.
     * Returns whether the whole file was handed off.
     */
    suspend fun sendFile(
        path: String,
        name: String,
        mime: String,
        id: String,
        send: suspend (Frame) -> Boolean,
    ): Boolean {
        val file = File(path)
        if (!send(Frame.FileStart(id, name, mime, file.length()))) return false
        val input = FileInputStream(file)
        try {
            val buffer = ByteArray(CHUNK)
            var seq = 0
            while (true) {
                val read = input.read(buffer)
                if (read <= 0) break
                val slice = if (read == buffer.size) buffer else buffer.copyOf(read)
                val encoded = Base64.encodeToString(slice, Base64.NO_WRAP)
                if (!send(Frame.FileData(id, seq++, encoded))) return false
            }
        } finally {
            input.close()
        }
        return send(Frame.FileEnd(id))
    }

    private fun downscale(bitmap: Bitmap, max: Int): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        if (width <= max && height <= max) return bitmap
        val ratio = minOf(max.toFloat() / width, max.toFloat() / height)
        return Bitmap.createScaledBitmap(bitmap, (width * ratio).toInt(), (height * ratio).toInt(), true)
    }

    private fun queryName(uri: Uri): String? {
        val cursor = appContext.contentResolver.query(
            uri,
            arrayOf(OpenableColumns.DISPLAY_NAME),
            null,
            null,
            null,
        ) ?: return null
        cursor.use {
            if (it.moveToFirst()) {
                val index = it.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                if (index >= 0) return it.getString(index)
            }
        }
        return null
    }

    private fun sanitize(name: String): String =
        name.replace(Regex("[^A-Za-z0-9._-]"), "_").take(80).ifEmpty { "file" }

    companion object {
        private const val CHUNK = 8 * 1024
        private const val MAX_IMAGE_DIMENSION = 1280
    }
}

File: ios/BlueTalk/Protocol/FileTransfer.swift
import Foundation
import UniformTypeIdentifiers
#if canImport(UIKit)
import UIKit
#endif

/// A file stored locally, ready to attach to a message or render.
struct FileAttachment {
    let path: String
    let name: String
    let mime: String
}

/// Moves files between devices over the chat channel, matching the Android
/// side: a file is sent as a `fileStart`, ordered `fileData` slices, and a
/// `fileEnd`, and streamed straight to disk on receipt. Outgoing images are
/// downscaled and re-encoded so a photo doesn't crawl over Bluetooth.
///
/// All calls happen on the main thread (the Bluetooth controllers deliver
/// their callbacks there), so no additional locking is needed.
final class FileTransfer {

    private let dir: URL

    init() {
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        dir = documents.appendingPathComponent("attachments")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    private struct Partial {
        let name: String
        let mime: String
        let url: URL
        let handle: FileHandle
    }

    private var partials: [String: Partial] = [:]

    // MARK: - Receiving

    func startIncoming(id: String, name: String, mime: String) {
        let url = dir.appendingPathComponent("\(id)-\(sanitized(name))")
        FileManager.default.createFile(atPath: url.path, contents: nil)
        guard let handle = try? FileHandle(forWritingTo: url) else { return }
        partials[id] = Partial(name: name, mime: mime, url: url, handle: handle)
    }

    func appendIncoming(id: String, base64: String) {
        guard let partial = partials[id], let data = Data(base64Encoded: base64) else { return }
        try? partial.handle.write(contentsOf: data)
    }

    func finishIncoming(id: String) -> FileAttachment? {
        guard let partial = partials.removeValue(forKey: id) else { return nil }
        try? partial.handle.close()
        return FileAttachment(path: partial.url.path, name: partial.name, mime: partial.mime)
    }

    func abortIncoming(id: String) {
        guard let partial = partials.removeValue(forKey: id) else { return }
        try? partial.handle.close()
        try? FileManager.default.removeItem(at: partial.url)
    }

    // MARK: - Sending

    /// Copies a picked file into app storage, downscaling and recompressing
    /// images. Returns nil if the file can't be read.
    func importOutgoing(from url: URL) -> FileAttachment? {
        let scoped = url.startAccessingSecurityScopedResource()
        defer { if scoped { url.stopAccessingSecurityScopedResource() } }

        let id = UUID().uuidString
        let mime = mimeType(for: url)
        let name = url.lastPathComponent

        #if canImport(UIKit)
        if mime.hasPrefix("image/"),
           let data = try? Data(contentsOf: url),
           let image = UIImage(data: data),
           let jpeg = downscale(image, max: 1280).jpegData(compressionQuality: 0.8) {
            let base = sanitized((name as NSString).deletingPathExtension)
            let dest = dir.appendingPathComponent("\(id)-\(base).jpg")
            guard (try? jpeg.write(to: dest)) != nil else { return nil }
            return FileAttachment(path: dest.path, name: dest.lastPathComponent, mime: "image/jpeg")
        }
        #endif

        let dest = dir.appendingPathComponent("\(id)-\(sanitized(name))")
        do {
            try FileManager.default.copyItem(at: url, to: dest)
        } catch {
            return nil
        }
        return FileAttachment(path: dest.path, name: name, mime: mime)
    }

    /// Streams a file to a peer as protocol frames, passing each to `send`.
    func sendFile(path: String, name: String, mime: String, id: String, send: (Frame) -> Void) {
        guard let handle = FileHandle(forReadingAtPath: path) else { return }
        defer { try? handle.close() }
        let attributes = try? FileManager.default.attributesOfItem(atPath: path)
        let size = (attributes?[.size] as? Int) ?? 0
        send(.fileStart(id: id, name: name, mime: mime, size: Int64(size)))
        var seq = 0
        while true {
            let chunk = handle.readData(ofLength: chunkSize)
            if chunk.isEmpty { break }
            send(.fileData(id: id, seq: seq, data: chunk.base64EncodedString()))
            seq += 1
        }
        send(.fileEnd(id: id))
    }

    // MARK: - Helpers

    private func mimeType(for url: URL) -> String {
        if let type = UTType(filenameExtension: url.pathExtension),
           let mime = type.preferredMIMEType {
            return mime
        }
        return "application/octet-stream"
    }

    #if canImport(UIKit)
    private func downscale(_ image: UIImage, max: CGFloat) -> UIImage {
        let width = image.size.width
        let height = image.size.height
        if width <= max && height <= max { return image }
        let ratio = min(max / width, max / height)
        let newSize = CGSize(width: width * ratio, height: height * ratio)
        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in image.draw(in: CGRect(origin: .zero, size: newSize)) }
    }
    #endif

    private func sanitized(_ name: String) -> String {
        let allowed = CharacterSet(charactersIn:
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789._-")
        let cleaned = String(name.unicodeScalars.map { allowed.contains($0) ? Character($0) : "_" })
        let trimmed = String(cleaned.prefix(80))
        return trimmed.isEmpty ? "file" : trimmed
    }

    private let chunkSize = 8 * 1024
}
